pkg/authorization/admission/restrictusers/subjectchecker.go

```diff
@@ -0,0 +1,181 @@
+
+	// groupSelectors is a list of label selectors.
+	groupSelectors []labels.Selector
+
+	// serviceaccountLiterals is a list of tuples of literal service account name
+	// and literal namespace name.
+	serviceaccountLiterals []*authorizationapi.ServiceAccountInNamespace
+
+	// serviceaccountInTheseNamespacesLiterals is a list of literal service
+	// account names.
+	serviceaccountInTheseNamespacesLiterals []string
+}
+
+// NewRoleBindingRestrictions returns a new RoleBindingRestrictions object with
+// no rules.
+func NewRoleBindingRestrictions(ns string, kc kclientset.Interface,
+	oc oclient.Interface) (*RoleBindingRestrictions, error) {
+	return &RoleBindingRestrictions{namespace: ns, kclient: kc, oclient: oc}, nil
+}
+
+// Allowed makes the determination whether a given subject is a collaborator on
+// the namespace.
+func (r *RoleBindingRestrictions) Allowed(subject *kapi.ObjectReference) error {
+	switch subject.Kind {
+	case authorizationapi.SystemUserKind:
+		return fmt.Errorf("system users are not allowed as collaborators: %q",
+			subject.Name)
+
+	case authorizationapi.SystemGroupKind:
+		return fmt.Errorf("system groups are not allowed as collaborators: %q",
+			subject.Name)
+
+	case authorizationapi.UserKind:
+		user, err := r.oclient.Users().Get(subject.Name)
+		if err != nil {
+			return err
+		}
+
+		for _, literal := range r.userLiterals {
+			if user.Name == literal {
+				return nil
+			}
+		}
+
+		for _, literal := range r.userInGroupLiterals {
+			for _, group := range user.Groups {
+				if group == literal {
+					return nil
+				}
+			}
+		}
+
+		for _, selector := range r.userSelectors {
+			if selector.Matches(labels.Set(user.Labels)) {
+				return nil
+			}
+		}
+	case authorizationapi.GroupKind:
+		group, err := r.oclient.Groups().Get(subject.Name)
+		if err != nil {
+			return err
+		}
+
+		for _, literal := range r.groupLiterals {
+			if group.Name == literal {
+				return nil
+			}
+		}
+
+		for _, selector := range r.groupSelectors {
+			if selector.Matches(labels.Set(group.Labels)) {
+				return nil
+			}
+		}
+	case authorizationapi.ServiceAccountKind:
+		sa, err := r.kclient.Core().ServiceAccounts(r.namespace).Get(subject.Name)
+		if err != nil {
+			return err
+		}
+
+		for _, literal := range r.serviceaccountLiterals {
+			if sa.Name == literal.Name && sa.Namespace == literal.Namespace {
+				return nil
+			}
+		}
+
+		for _, literal := range r.serviceaccountInTheseNamespacesLiterals {
+			if sa.Namespace == literal {
+				return nil
+			}
+		}
+	}
+
+	return fmt.Errorf("%s %q is not a collaborator on project %q",
+		subject.Kind, subject.Name, r.namespace)
+}
+
+// AddMatcher adds rules to the RoleBindingRestrictions object based on the
+// given CollaboratorMatcher object.
+func (r *RoleBindingRestrictions) AddMatcher(cm *authorizationapi.CollaboratorMatcher) error {
+	switch {
+	case cm.Spec.UserMatcher != nil:
+		r.userLiterals = append(r.userLiterals, cm.Spec.UserMatcher.Users...)
```

Comment:
This is exactly the kind of code I want to avoid.  Building up a structure like this adds a layer that I don't think we need.